### ISSUE TITLE
Restrict SimpleSnitch to work with SimpleStrategy only and vise versa

### ISSUE THREAD

**Issue by vladzcloudius** [scylladb/scylla#1789]
Currently there are absolutely no restrictions on a replication strategy that may be used during the keyspace creation. However using a SimpleSnitch or a SimpleStrategy with some DC-aware components may (and does) create some unexpected failures at customer sites. 

Although DataStax does not explicitly prohibit using SimpleXXX components with non-Simple ones it do clearly states that they should only be used together or not used together:

["CREATE KEYSPACE" doc](https://docs.datastax.com/en/cql/3.1/cql/cql_reference/create_keyspace_r.html): "Using SimpleStrategy is fine for evaluating Cassandra. For production use or for use with mixed workloads, use NetworkTopologyStrategy."
[Simple Snitch doc](https://docs.datastax.com/en/cassandra/3.x/cassandra/architecture/archSnitchSimple.html): "Using a SimpleSnitch, you define the keyspace to use SimpleStrategy and specify a replication factor."  

However apparently not all users (and some times developers) read the documentation. We have faced a few times the situations when users mixed SimpleXXX components with DC-aware ones and things didn't work as they expected. 

Having said all that it makes sense to add a check that would prevent creating a keyspace with a non-Simple strategy is a snitch is a SimpleSnitch or creating a keyspace with a SimpleStrategy when a snitch is a non-Simple one.

**Comment by nyh** [comment]
Hi @vladzcloudius in my fix for #6635 I added an option (which is still not the default, but can be) `--restrict-replication-simplestrategy true` which forbids creating keyspaces with SimpleStrategy.

My question is now what, if anything, is left to be done in this issue. 

In the title of the issue, you suggested that the Simple-ness of the snitch should match the Simple-ness of the repliation strategy. But if I understand correctly, even if we have just one DC, we still want people to use `NetworkTopologyStrategy`, in case another DC is added later.

**Comment by vladzcloudius** [comment]
@nyh Requiring people to use `NetworkingTopologyStrategy` (what you did) and forbidding using a `SimpleSnitch` with anything but a `SimpleStrategy` is not the same thing. - but we want to be able to have both those restrictions. 

The `SimpleSnitch` + `NetworkingStrategy` combo is harmful too because `NetworkingStrategy` is not only about multi-DC but is also about rack-awarness and a `SimpleSnitch` kills that enforcing a single rack. 

The bottom line: if we would have both your and this GH restrictions enabled we would enforce both NetworkingStrategy *and* non-simple Snitch - which is exactly the configuration we want to have on every production installation.

**Comment by nyh** [comment]
@vladzcloudius you'll have to explain this slowly to me because I don't know anything about how snitches are configured.
Am I correct that you don't choose a snitch when creating a keyspace, but rather it's a per-node (why?) setting in scylla.yaml?

If this is the case, then what do you propose should happen when one boots with SimpleSnitch which is the default? Are you saying that in that case (if the restriction is turned on) I should forbid creating a keyspace with NetworkTopologyStrategy? This is the opposite of the restriction I already added, which is that you should **only** use NetworkTopologyStrategy?

Another alternative, which sounds better to me but sounds like is *not* what you want, is that perhaps we shouldn't allow booting Scylla if it is configured to have multiple racks or dcs but using SimpleSnitch. Can you please tell me what's wrong with this alternative - which is also simpler because it doesn't mix the snitch and replication strategy into one logic?

**Comment by vladzcloudius** [comment]
> @vladzcloudius you'll have to explain this slowly to me because I don't know anything about how snitches are configured.

No problem at all, sir! :)

> Am I correct that you don't choose a snitch when creating a keyspace, but rather it's a per-node (why?) setting in scylla.yaml?

That's correct.
Why? - This how it is in Cassandra and we decided not to change it. Probably, because it would make migrations more complicated and we would no longer be a "drop-in Cassandra replacement" ;)

> 
> If this is the case, then what do you propose should happen when one boots with SimpleSnitch which is the default? Are you saying that in that case (if the restriction is turned on) I should forbid creating a keyspace with NetworkTopologyStrategy? 

Correct. This should shake a user up and make him read the corresponding documentation that clearly states that a SimpleSnitch *should never be used* is a production system.

> This is the opposite of the restriction I already added, which is that you should **only** use NetworkTopologyStrategy?

I wouldn't call it the opposite.
I think that your restriction is a subset of my restriction because it makes little sense to use NetworkingStrategy with a Simple snitch as I explained earlier.

Hence if a SimpleSnitch *IS* configured at least get some notification that using a NetworkingStrategy is not going to give you the benefit you were hoping for.

Users usually understand very well why they need to use NetworkingTopologyStrategy the moment they realize that it exists. The same holds to using non-simple Snitch.

And if some user insists on using Simple stuff - he/she can always override those restrictions.

However the majority of users are going to switch to the non-Simple configurations once they see errors triggered by the corresponding restriction.

> 
> Another alternative, which sounds better to me but sounds like is _not_ what you want, is that perhaps we shouldn't allow booting Scylla if it is configured to have multiple racks or dcs but using SimpleSnitch. Can you please tell me what's wrong with this alternative - which is also simpler because it doesn't mix the snitch and replication strategy into one logic?

The problem with the "alternative" above is that the Snitch is currently the primary "source of truth" which defines if there are multiple DCs/RACKs and what are they. And if a SimpleSnitch is used it will always return a single DC and a single RACK for any cluster.

What we could have done however is to implement a restriction similar to what you have done for NetworkingStrategy and forbid the SimpleSnitch if the corresponding (new) option is enabled.

However this GH proposal allows to have both those restriction in a "single package" which makes more sense since both Snitch and NetworkingTopology restriction should always be on or off together.

**Comment by nyh** [comment]
@vladzcloudius I still don't follow your logic on tying the two things (simplesnitch and simplestrategy) together:

We already decided that `NetworkToplogyStrategy` should be recommended **always**, regardless of whether or not you actuall have racks or DCs. It can be used even one a one-node cluster, and this is what we do in tests (or will do - see #8638 ).

I don't want to have a case (when SimpleSnitch is used) where we reverse this recommendation and say that now you must use SimpleStrategy. No. I want NetworkTopologyStrategy to be used in that case too - even if it makes no difference.

You can tell me now it was wrong to use SimpleSnitch. Ok. So let's forbid SimpleSnitch. But it would be just a check on the snitch - and not a combination of snitch and replication strategy.

If you agree with me that the restriction should be "don't allow simplesnitch" I'll close this issue and open one to replace it which talks just about restricting the allowed snitch.

**Comment by vladzcloudius** [comment]
@nyh I'm coming from the point where we DO support both `SimpleSnitch` and `SimpleStrategy`.

If we want to stop supporting them - that would make all this discussion and corresponding new options/restrictions not necessary. 

And as long as we DO support them I believe that this proposal makes more sense than having a separate restriction for a Snitch and a Strategy.

**Comment by nyh** [comment]
We already decided (@slivne, @eliransin were involved in this decision) that we will stop supporting SimpleStrategy completely, it's just that it will happen in stages - first we already added an option (restrict_replication_simple_strategy) that forbids SimpleStrategy. This option is not yet the default, but already specific installations (e.g., Scylla Cloud) can enable it, and eventually it will become the default and users will need to explicitly disable it if they don't want that protection.

What your proposal means is that if somebody booted with `SimpleSnitch` he won't be told anything is wrong until trying to create a table, and at that point he will be told he **must** create a table with `SimpleStrategy`, which is the opposite of what we want him to do.

I want to understand something: Is there any **benefit** to booting with SimpleSnitch? Is it somehow more efficient when you don't have DCs or racks? If there is no benefit, why should we allow SimpleSnitch at all? (of course when I say "don't allow" it can be overriden, but by default, shouldn't we forbid it?)

**Comment by vladzcloudius** [comment]
@nyh If we want to keep allowing booting with a SimpleSnitch at all (with whatever options it requires) and if we want to continue allowing using a SimpleStrategy at all (with whatever options it requires) then my proposal makes sense for reasons I described earlier.

> I want to understand something: Is there any benefit to booting with SimpleSnitch? Is it somehow more efficient when you don't have DCs or racks? If there is no benefit, why should we allow SimpleSnitch at all? (of course when I say "don't allow" it can be overriden, but by default, shouldn't we forbid it?)

The only reason for an existence of a `SimpleSnitch` and a `SimpleStrategy` is a simplicity (dah): both require no additional configuration (e.g. defining DC/Racks configs or having meta-servers available like with Ec2Snitch). 

Can people be off without them - probably, yes (e.g. use `GossipingPropertyFileSnitch`, some default values in `cassandra-rackdc.properties` and `NetworkingStrategy` that would use it). 

It's not clear to me however how we would guarantee the robustness of the system when it comes to internal distributed KSes like `system_auth` and `system_distributed` which use `SimpleStrategy` by default today. The advantage of SimpleStrategy as a default for these is that it would make the data of those to be replicated no matter what is the content of all other configs while with non-Simple config you can find yourself with a zero-replication state (e.g. when the DC name in the KS replication config and in the Snitch config do not match).

But I believe @slivne and @eliransin have some ideas about that.

On top of all that and assuming we ban `SimpleXX` by default at some point as you wrote, somebody can still prefer the simplicity of `SimpleXXX` in some circumstances (e.g. K8S, docker, etc.) and enable them. In such cases you don't want to allow turning the `SimpleSnitch` on if there are already existing KSes with a `NetworkingStrategy`  or create new KSes with a `NetworkingStrategy` if a `SimpleSnitch` is used - which is what this issue is about.

**Comment by eliransin** [comment]
IMO if we are still to allow SimpleStrategy in any constellation we should protect against the snitch mismatch. My justification is 
that those restrictions are there to help the user not to shoot himself in the foot but lets say he decided to drop one restriction we 
still don't want him to make mistakes, if this is not the case it is equivalent to having one big restriction for strategy (lets call it 
`strategy_restrictions` for the sake of argument) and allowing the user to get this protection or not but nothing in between.
What I do agree that this might be a lower priority restriction, at least starting from the day the simple_strategy restriction will be 
on by default since we expect it to reduce significantly the amount of users that use it.
One last remark about this, I think that the log should refer the user to the documentation that explains the reasoning behind this
so he can make an educated decisions and not just turning the restrictions off.
